elk/training/train.py

```diff
@@ -54,114 +127,64 @@ def train(args):
-        val_x0, val_x1 = val_h.to(args.device).float().chunk(2, dim=-1)
-
-        train_labels_aug = torch.cat([train_labels, 1 - train_labels])
-        val_labels_aug = torch.cat([val_labels, 1 - val_labels])
-        if pbar:
-            pbar.set_description("Fitting CCS")
-
-        # Instantiate the reporter.
-        reporter = Reporter(
-            in_features=x0.shape[-1],
-            device=args.device,
-            init=args.init,
-            loss=args.loss,
-            supervised_weight=args.supervised_weight,
-        )
-        if args.label_frac:
-            num_labels = round(args.label_frac * len(train_labels))
-            labels = train_labels[:num_labels].to(args.device)
-        else:
-            labels = None
-
-        train_loss = reporter.fit(
-            contrast_pair=(x0, x1),
-            labels=labels,
-            num_tries=args.num_tries,
-            optimizer=args.optimizer,
-            weight_decay=args.weight_decay,
-        )
-        val_result = reporter.score(
-            (val_x0, val_x1),
-            val_labels.to(args.device),
-        )
-        if pbar:
-            pbar.set_postfix(train_loss=train_loss, ccs_auroc=val_result.auroc)
-        stats = [train_loss, *val_result]
-
-        if not args.skip_baseline and not dist.is_initialized():
-            if pbar:
-                pbar.set_description("Fitting LR")
-
-            # TODO: Once we implement cross-validation for CCS, we should benchmark
-            # against LogisticRegressionCV here.
-            lr_model = LogisticRegression(max_iter=10_000)
-            lr_model.fit(torch.cat([x0, x1]).cpu(), train_labels_aug)
-
-            lr_preds = lr_model.predict_proba(torch.cat([val_x0, val_x1]).cpu())[:, 1]
-            lr_acc = accuracy_score(val_labels_aug, lr_preds > 0.5)
-            lr_auroc = roc_auc_score(val_labels_aug, lr_preds)
-            if pbar:
-                pbar.set_postfix(
-                    train_loss=train_loss, ccs_auroc=val_result.auroc, lr_auroc=lr_auroc
-                )
-
-            lr_models.append(lr_model)
-            stats += [lr_auroc, lr_acc]
-
-        statistics.append(stats)
-        reporters.append(reporter)
-
-    reporters.reverse()
-    lr_models.reverse()
 
-    path = elk_cache_dir() / args.name
-    if rank == 0:
+    # Collect the results and update the progress bar.
+    with open(cache_dir / "eval.csv", "w") as f:
         cols = ["layer", "train_loss", "loss", "acc", "cal_acc", "auroc"]
         if not args.skip_baseline:
             cols += ["lr_auroc", "lr_acc"]
 
-        with open(cache_dir / "eval.csv", "w") as f:
-            writer = csv.writer(f)
-            writer.writerow(cols)
+        writer = csv.writer(f)
+        writer.writerow(cols)
+
+        # Don't bother with multiprocessing if we're only using one device.
+        if args.num_devices == 1:
+            pbar = tqdm(iterator, total=L, unit="layer")
+            for i, (train_h, val_h) in enumerate(pbar):
```

Comment:
Currently this claims that the layer is the same thing as the index into train_h and val_h and like you said this is broken. It might be good to flag that. This will be redone in versions where the layer information is available.

Reply:
Added a warning in `elk extract` when you try to save only a subset of layers as a bandaid solution